Spark MicroBatch read - Ignore replace snapshots and add Spark option to skip delete snapshots

daksha121 opened the pull request:
Building on top of #2660, this PR introduces the ability to skip deletes in Micro Batch read path. It also ignores replace snapshots.
Spark option introduced:
- To skip deletes: `streaming-skip-delete-snapshots`
Additional contributor to this PR: @SreeramGarlapati

SreeramGarlapati commented on spark/src/main/java/org/apache/iceberg/spark/SparkReadOptions.java:
```diff
@@ -50,4 +50,10 @@ private SparkReadOptions() {
   public static final String FILE_SCAN_TASK_SET_ID = "file-scan-task-set-id";
+
+  // skip snapshots of type delete while reading stream out of iceberg table
+  public static final String READ_STREAM_SKIP_DELETE = "read-stream-skip-delete";
```
on a second thought this naming seems like a better one - `skip-deletes-on-stream-read`
my problem with the older one was that - `skipping deletes` is not popping out - `read-stream` is.

rdblue commented on spark/src/main/java/org/apache/iceberg/spark/SparkReadOptions.java:
```diff
@@ -50,4 +50,10 @@ private SparkReadOptions() {
+  public static final String READ_STREAM_SKIP_DELETE = "read-stream-skip-delete";
+
+  // skip snapshots of type replace while reading stream out of iceberg table
+  public static final String READ_STREAM_SKIP_REPLACE = "read-stream-skip-replace";
```
Why make this optional? Since replace commits don't affect the logical data in the table, shouldn't they always be skipped?

daksha121 replied:
That makes sense. Added an explicit option based off of my interpretation of https://github.com/apache/iceberg/pull/2660#discussion_r658351562
But maybe an explicit option was not the intention of the comment? :)

rdblue replied:
No, I wasn't implying that there should be an option. I was saying they should just be skipped.

rdblue commented on spark3/src/test/java/org/apache/iceberg/spark/source/TestStructuredStreamingRead3.java:
```diff
@@ -363,6 +390,45 @@ public void testReadStreamWithSnapshotTypeDeleteErrorsOut() throws Exception {
+  @Test
+  public void testReadStreamWithSnapshotTypeDeleteAndSkipDeleteOption() throws Exception {
+    table.updateSpec()
+            .removeField("id_bucket")
```
Indentation is off in this method.

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java:
```diff
@@ -169,35 +174,64 @@ public void stop() {
-    Snapshot snapshotAfter = SnapshotUtil.snapshotAfter(table, snapshotId);
+  private boolean shouldSkip(Snapshot snapshot) {
+    if (snapshot.operation().equals(DataOperations.DELETE)) {
+      return shouldSkipDelete(snapshot);
```
I think it would be cleaner to embed the `shouldSkip` methods here rather than having a method that is a precondition and a static `return true`:

```java
  private boolean shouldProcess(Snapshot snapshot) {
    DataOperation op = snapshot.operation();
    Preconditions.checkState(op != DataOperations.DELETE || skipDelete,
        "Cannot process delete snapshot: %s", snapshot.snapshotId());
    Preconditions.checkState(op == DataOperations.APPEND || op == DataOperations.REPLACE,
        "Cannot process %s snapshot: %s", op.toString.toLower(Locale.ROOT), snapshot.snapshotId());
    return op == DataOperations.APPEND;
  }
```

I also used `shouldProcess` instead of `shouldSkip` to make it easier to read.

rdblue commented on spark3/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java:
```diff
@@ -169,35 +174,64 @@ public void stop() {
+
+    Preconditions.checkState(
+        snapshot.operation().equals(DataOperations.APPEND),
+        "Invalid Snapshot operation: %s, only APPEND is allowed.", snapshot.operation());
```
I think these error messages could be a bit more clear. The problem is that the snapshot cannot be processed because it is a replace. So I would use "Cannot process replace snapshot". Also, "only APPEND is allowed" is not true so I think this should probably not state it.

RussellSpitzer commented on spark3/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java:
```diff
@@ -169,35 +173,44 @@ public void stop() {
-          latestMicroBatch != null && latestMicroBatch.lastIndexOfSnapshot() ?
-              new StreamingOffset(snapshotAfter(latestMicroBatch.snapshotId()), 0L, false) :
-              batchStartOffset;
+      if (currentOffset == null) {
```
Does this mean that we always process batchStartOffsets? Because couldn't our oldest offset be in a delete snapshot?

daksha121 replied:
My bad, will fix in the next commit

RussellSpitzer commented on spark3/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java:
```diff
@@ -169,35 +173,44 @@ public void stop() {
+      } else {
+        Snapshot snapshotAfter = SnapshotUtil.snapshotAfter(table, currentOffset.snapshotId());
+        boolean shouldProcess = shouldProcess(snapshotAfter);
+        currentOffset = new StreamingOffset(snapshotAfter.snapshotId(), 0L, false);
```
I wonder if this bit of code would be easier to read if it was broken up into

"Set currentOffset"

"Decided if shouldProcess CurrentOffset?"

Does that make sense?

RussellSpitzer commented on spark3/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java:
```diff
@@ -169,35 +173,44 @@ public void stop() {
+        currentOffset = new StreamingOffset(snapshotAfter.snapshotId(), 0L, false);
+
+        if (!shouldProcess) {
+          continue;
```
I think we may want to log (at least at debug but maybe info) that we are skipping a snapshot

RussellSpitzer reviewed (commented):
Looks good to me, I had a few questions on the skip/current offset logic. Let me know when I should take another look

daksha121 commented:
> Looks good to me, I had a few questions on the skip/current offset logic. Let me know when I should take another look

Thanks for the review @RussellSpitzer! I addressed the comments and it's ready for another look

RussellSpitzer commented on spark3/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java:
```diff
@@ -169,35 +176,44 @@ public void stop() {
 
-      latestMicroBatch = MicroBatches.from(table.snapshot(currentOffset.snapshotId()), table.io())
+      if(!shouldProcess(table.snapshot(currentOffset.snapshotId()))) {
+        LOG.debug("Skipping snapshot: {}", currentOffset.snapshotId());
```
Probably add a little more info about the context of this skip, Skipping snapshot in stream x? of table y? Whatever you feel is appropriate

RussellSpitzer reviewed (approved):
Had tiny minor comment on debug method but this looks good to me

RussellSpitzer commented:
Hitting some style check failures 
```
Error: eckstyle] [ERROR] /home/runner/work/iceberg/iceberg/spark3/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java:194:7: 'if' is not followed by whitespace. [WhitespaceAfter]
```

Be sure to either enable the style checker in your IDE or run the gradle based tests
```
g build -x test  -x integrationTest
```

daksha121 commented:
Addressed your comments, @rdblue this is ready for another pass. Thanks!

RussellSpitzer commented:
@daksha121 I think we are still waiting on a change to the parameter name. I'm not good on names so I generally am fine with almost anything but I believe we are still hoping for a better name for the parameter

daksha121 commented:
> @daksha121 I think we are still waiting on a change to the parameter name. I'm not good on names so I generally am fine with almost anything but I believe we are still hoping for a better name for the parameter

I'm not clear on which parameter needs to be renamed.. Are you referring to the new spark option (skip-deletes-on-stream-read) that this PR introduces?

daksha121 commented:
> > @daksha121 I think we are still waiting on a change to the parameter name. I'm not good on names so I generally am fine with almost anything but I believe we are still hoping for a better name for the parameter
> 
> I'm not clear on which parameter needs to be renamed.. Are you referring to the new spark option (skip-deletes-on-stream-read) that this PR introduces?

@RussellSpitzer how about just:
- skip-deletes OR
- ignore-deletes

RussellSpitzer commented:
@rdblue what do you think? I'm fine with `streaming.ignore-deletes` or anything like that.

rdblue commented:
We try not to use namespaces in the read options to keep them simple. This one is a bit odd because `ignore-deletes` could be misinterpreted easily -- we don't want anyone to think they can read a table without applying delete files, for example. So I want to make it clear that this is streaming and that this not delete files. The best I'm coming up with is `skip-delete-snapshots`, which is clear that it is referring to snapshot operations. I don't think it would be misinterpreted if it were used in a batch context since it wouldn't make sense to run a select that doesn't do anything if the latest snapshot is a delete.

daksha121 commented:
> We try not to use namespaces in the read options to keep them simple. This one is a bit odd because `ignore-deletes` could be misinterpreted easily -- we don't want anyone to think they can read a table without applying delete files, for example. So I want to make it clear that this is streaming and that this not delete files. The best I'm coming up with is `skip-delete-snapshots`, which is clear that it is referring to snapshot operations. I don't think it would be misinterpreted if it were used in a batch context since it wouldn't make sense to run a select that doesn't do anything if the latest snapshot is a delete.

Thanks @rdblue, that makes sense. Can you help understand why we wouldn't want to put the word streaming in like `streaming-skip-deletes`? We could perhaps eliminate the risk of misinterpretation. Maybe I'm missing something here

rdblue commented:
I'd be fine adding `streaming-`. I was just trying to keep it as small as possible. We definitely need `-snapshots` because it should be clear that this doesn't affect row-level deletes.

daksha121 commented:
> I'd be fine adding `streaming-`. I was just trying to keep it as small as possible. We definitely need `-snapshots` because it should be clear that this doesn't affect row-level deletes.

Thanks @rdblue. Renamed the option to `streaming-skip-delete-snapshots`

SreeramGarlapati commented:
@rdblue - pl. drop me as a contributor for this work. @daksha121 - was nice enough to mention me as one. I was trying to pay back her help in work: #2660.

rdblue commented:
Thanks, @daksha121! I merged this.
